manuals/tableauinfo.1
.Dd September  7, 2026
.Dt TABLEAUINFO 1
.Os
.Sh NAME
.Nm tableauinfo
.Nd determines information about a Tableau write blocker
.Sh SYNOPSIS
.Nm tableauinfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm tableauinfo
is a utility to determine information about a Tableau write blocker
.Pp
.Nm tableauinfo
is part of the
.Nm libtableau
package.
.Nm libtableau
is a library to support the Tableau forensic bridge (write blocker)
.Pp
.Ar source
is the source device.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# tableauinfo /dev/sdg
tableauinfo 20080405 (libtableau 20080405, libsgutils 1.32 20070129)
.sp
Tableau forensic bridge information
        Vendor:                 Tableau
        Model:                  T5
        Serial number:          0123456789abcdef
        Firmware date:          Feb 22 2008
        Firmware time:          10:58:48
        Firmware stepping:      4
        Debug firmware:         false
        Channel index:          0
        Channel type:           PATA
        Read only mode:         true
        Read only reporting:    true
        Write error reporting:  true
.sp
Drive information
        Vendor:
        Model:                  WDC AC22100L
        Serial number:          WD-WM3456789123
        Revision number:        32.41N37
.sp
Drive security information
        HPA support:            false
        HPA in use:             false
        DCO support:            false
        DCO in use:             false
        Security support:       false
        Security in use:        false
        Amount of sectors:      4124736
        HPA amount of sectors:  4124736
        DCO amount of sectors:  4124736
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr tableaucontrol 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libtableau/issues
.Sh COPYRIGHT
Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* improve error handling for send command
  - detect response and no response errors
* create debian package files

20170903
* Applied updates
* see `git log' for more recent change log

20160819
* Applied updates

20080511
* Fixed typo in spec file
* added features include header
* Updated email address

20080502
* Fixed .pc file
* Fixed make lib

20080426
* Minor changes

20080416
* Minor changes

20080404
* Worked on manuals
* Worked on initial version

20080404
* Worked on initial version

20080403
* Added initial version of MacOS-X project files

20080329
* Worked on initial version
* Licensed code under LGPL v3
* Created platform independent SCSI IO using libsgutils
* Added print of libsgutils version in tableauinfo and control output

20080325
* Worked on initial version

20080324
* Worked on initial version

20080323
* Worked on initial version

20080322
* Worked on initial version
* integrated libewf common code base

20080122
* Worked on initial version

20080102
* Created initial version based on tableau-parm-0.1.0.tar.gz



tableautools/tableauinfo.c
/*
 * Shows information obtained from a Tableau write blocker.
 *
 * Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "tableautools_getopt.h"
#include "tableautools_libcerror.h"
#include "tableautools_libclocale.h"
#include "tableautools_libcnotify.h"
#include "tableautools_libtableau.h"
#include "tableautools_output.h"
#include "tableautools_signal.h"
#include "tableautools_unused.h"

info_handle_t *tableauinfo_info_handle = NULL;
int tableauinfo_abort                  = 0;

/* Signal handler for tableauinfo
 */
void tableauinfo_signal_handler(
      tableautools_signal_t signal TABLEAUTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "tableauinfo_signal_handler";

	TABLEAUTOOLS_UNREFERENCED_PARAMETER( signal )

	tableauinfo_abort = 1;

	if( tableauinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     tableauinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description    = \
		"Use tableauinfo to determine information about a Tableau write blocker.";

	tableautools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source device" },
	};
	system_character_t options_string[ 32 ];

	libtableau_error_t *error  = NULL;
	system_character_t *source = NULL;
	char *program              = "tableauinfo";
	system_integer_t option    = 0;
	int number_of_options      = (int) ( sizeof( options ) / sizeof( tableautools_option_t ) );
	int verbose                = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "tableautools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( tableautools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	tableautools_output_version_fprint(
	 stdout,
	 program );

	if( tableautools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = tableautools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				tableautools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				tableautools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				tableautools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source device.\n" );

		tableautools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libtableau_notify_set_stream(
	 stderr,
	 NULL );
	libtableau_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &tableauinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open_input(
	     tableauinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source device.\n" );

		goto on_error;
	}
	if( info_handle_device_fprint(
	     tableauinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print device information.\n" );

		goto on_error;
	}
	if( info_handle_close_input(
	     tableauinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &tableauinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( tableauinfo_info_handle != NULL )
	{
		info_handle_free(
		 &tableauinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

